logstash-core/lib/logstash/agent.rb

```diff
@@ -10,80 +10,125 @@
-    begin
-      # There should be a better way to test this ideally
-      LogStash::Pipeline.new(config_str)
-    rescue Exception => e
-      e
+  def execute
+    @thread = Thread.current
+    @logger.info("starting agent", :state => @state)
+
+    if @auto_reload
+      Stud.interval(@reload_interval) do
+        break unless clean_state? || running_pipelines?
+        reload_state!
+      end
+    else
+      reload_state!
+      while !Stud.stop?
+        break unless clean_state? || running_pipelines?
+        sleep 0.5
+      end
     end
   end
 
-  def execute
-    # Make SIGINT/SIGTERM shutdown the pipeline.
-    sigint_id = trap_sigint()
-    sigterm_id = trap_sigterm()
+  def shutdown
+    shutdown_pipelines
+  end
 
-    @pipelines.each {|_, p| p.run } # blocking operation. works now because size <= 1
-    return 0
+  def reload_state!
+    new_state = fetch_state
+    if valid_state?(new_state)
+      if new_state?(@state, new_state)
+        @logger.warn("fetched new state. upgrading..", :state => new_state)
+        upgrade_state(new_state)
+      else
+        @logger.debug("same state, ignoring..")
+      end
+    else
+      @logger.error("invalid state", :state => new_state)
+    end
   rescue => e
     @logger.fatal I18n.t("oops", :error => e)
     @logger.fatal e.backtrace if @logger.debug? || $DEBUGLIST.include?("stacktrace")
     return 1
-  ensure
-    Stud::untrap("INT", sigint_id) unless sigint_id.nil?
-    Stud::untrap("TERM", sigterm_id) unless sigterm_id.nil?
-  end # def execute
+  end
 
+  private
   def add_pipeline(pipeline_id, config_str, settings = {})
     @pipelines[pipeline_id] = LogStash::Pipeline.new(config_str, settings)
   end
 
-  private
-  # Emit a warning message.
-  def warn(message)
-    # For now, all warnings are fatal.
-    raise LogStash::ConfigurationError, message
-  end # def warn
+  def start_pipeline(id)
+    return unless @pipelines[id]
+    @logger.info("starting pipeline", :id => id)
+    @pipeline_threads[id] = Thread.new do
+      LogStash::Util.set_thread_name("pipeline.#{id}")
+      @pipelines[id].run
+    end
+  end
 
-  # Emit a failure message and abort.
-  def fail(message)
-    raise LogStash::ConfigurationError, message
-  end # def fail
+  def stop_pipeline(id)
+    return unless @pipelines[id]
+    @logger.warn("stopping pipeline", :id => id)
+    @pipelines[id].shutdown do
+      LogStash::ShutdownController.start(@pipelines[id], @pipeline_threads[id])
+    end
+    @pipeline_threads[id].join
+  end
 
   def shutdown_pipelines
-    @pipelines.each do |_, pipeline|
-      pipeline.shutdown do
-        ::LogStash::ShutdownController.start(pipeline)
-      end
-    end
+    @pipelines.each { |id, _| stop_pipeline(id) }
   end
 
-  def trap_sigterm
-    Stud::trap("TERM") do
-      @logger.warn(I18n.t("logstash.agent.sigterm"))
-      shutdown_pipelines
-    end
+  def running_pipelines?
+    @pipeline_threads.select {|_, pipeline| pipeline.alive? }.any?
```

Comment:
Maybe call `.values` instead of using `_`. Totally optional, I'll leave it to you to make the choice here.